app/frontend/wallet/address-manager.js

```diff
@@ -9,21 +9,81 @@ const AddressManager = ({
   derivationScheme,
   disableCaching, // good for tests
   isChange,
+  blockchainExplorer,
 }) => {
   const state = {
     deriveAddressMemo: {},
   }
 
-  function discoverAddresses() {
+  async function discoverAddresses() {
+    switch (derivationScheme.type) {
+      case 'v1':
+        return await discoverAddressesV1()
+      case 'v2':
+        return await discoverAddressesV2()
+      default:
+        return null
+    }
+  }
+
+  /*
+   * AdaLite support discovery of only first 10 addresses by default (v1 scheme).
+   */
+  function discoverAddressesV1() {
     const childIndexBegin = derivationScheme.startAddressIndex
     const childIndexEnd = childIndexBegin + addressLimit
+    return discoverAddressesBlock(childIndexBegin, childIndexEnd)
+  }
+
+  /*
+   * To achieve compatibility with Yoroi wallet, we have to discover all addresses
+   * until gap limit is reached. We discover addresses by blocks of 20. This should
+   * be more foolproof (v2 scheme).
+   */
+  async function discoverAddressesV2() {
+    const blockSize = 20
+    let addresses = []
+    let isGapBlock = false
+    let childIndexBegin = derivationScheme.startAddressIndex
+    let blockItterator = 0
+    let childIndexEnd, block
+
+    while (!isGapBlock) {
+      childIndexEnd = childIndexBegin + blockSize
+      block = await discoverAddressesBlock(childIndexBegin, childIndexEnd)
+
+      isGapBlock = !(await blockchainExplorer.isSomeAddressUsed(block))
+      if (!isGapBlock || blockItterator === 0) addresses = [...addresses, ...block]
+
+      childIndexBegin += blockSize
+      blockItterator++
+    }
+    return await reduceAddresses(addresses)
+  }
+
+  /*
+   * Function which discovers addresses from begin index to end index
```

Comment:
this comment is redundant, and I would rather name the function to "getAbsDerivationPaths(...)" and return only the absolute derivation paths, since that's the trickiest part which is worth encapsulating.